File: fly/src/main/java/com/xia/fly/mvp/nullobject/NoOp.java
package com.xia.fly.mvp.nullobject;

import java.lang.reflect.InvocationHandler;
import java.lang.reflect.Method;

import static java.lang.reflect.Proxy.newProxyInstance;

/**
 * @author xia
 * @date 2018/7/19.
 */
public final class NoOp {

    private static final InvocationHandler DEFAULT_VALUE = new DefaultValueInvocationHandler();

    @SuppressWarnings("unchecked")
    public static <T> T of(Class<T> interfaceClass) {
        return (T) newProxyInstance(interfaceClass.getClassLoader(),
                new Class[]{interfaceClass},
                DEFAULT_VALUE);
    }

    private static class DefaultValueInvocationHandler implements InvocationHandler {
        @Override
        public Object invoke(Object proxy, Method method, Object[] args) {
            return Defaults.defaultValue(method.getReturnType());
        }
    }
}

File: fly/src/main/java/com/xia/fly/mvp/nullobject/NoOp.kt
package com.xia.fly.mvp.nullobject

import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy.newProxyInstance

/**
 * @author weixia
 * @date 2019/6/12.
 */
object NoOp {

    private val DEFAULT_VALUE = DefaultValueInvocationHandler()

    @Suppress("UNCHECKED_CAST")
    fun <T> of(interfaceClass: Class<T>): T {
        return newProxyInstance(interfaceClass.classLoader,
                arrayOf<Class<*>>(interfaceClass),
                DEFAULT_VALUE) as T
    }

    private class DefaultValueInvocationHandler : InvocationHandler {
        override fun invoke(proxy: Any?, method: Method?, args: Array<Any>?): Any? {
            return Defaults.defaultValue(method?.returnType)
        }
    }
}
